### PR TITLE
Deprecate types.loaOf

### PR DESCRIPTION
**Pull request by rnhmjoj**
###### Motivation for this change
Fix issue #1800.

It turns out this is a lot of work: there are many instances of lists being used instead of sets.
The main culprits are `users.users` and `environment.etc`.
I think I have identified most of them but there are still some, moreover all this changes should be properly tested so I definitely need some help.

### CODE REVIEW

File: nixos/modules/services/mail/exim.nix
Comment (Infinisil) [inline nixos/modules/services/mail/exim.nix]: Happen to see a mistake here

**Comment by Infinisil** [comment]
If we want to deprecate this, there needs to be some deprecation warning for users. We can't just flip the switch and possibly break every second person's config. A good way to do that is to change the `loaOf` type to issue a warning when the list case is used, something like

```
Warning: You're assigning the list [ { name = "foo"; bar = <CODE>; ... } ... ] to
<option name>, which won't be supported in the future, use an attribute set like
{ foo = { bar = <CODE>; ... }; ... } instead. See
https://github.com/NixOS/nixpkgs/pull/63103 for more info.
```

File: nixos/modules/services/mail/exim.nix
Comment (rnhmjoj) [reply nixos/modules/services/mail/exim.nix]: Thank you, fixed.

**Comment by rnhmjoj** [comment]
Ok, I removed the attempts to replace loaOf with attrsOf and added a deprecation warning.
I don't know how to do something like that though

File: lib/types.nix
Comment (Infinisil) [inline lib/types.nix]: Yeah this isn't what we need. The warning needs to go in the `merge` function. There one has access to all definitions, which can then be filtered to lists only, and for each of them one can print a warning as I showed in my comment. would be even neater to include the file and line where the user needs to do this change.

I generally think our error messages suck too much. It's not too difficult to make them better, especially with all the info the module system gives access to.

Comment (rnhmjoj) [reply lib/types.nix]: Ok, I'm working on it.

**Comment by rycee** [comment]
This will break the [programs.ssh.matchBlocks](https://rycee.gitlab.io/home-manager/options.html#opt-programs.ssh.matchBlocks) option in Home Manager. That option allows use of list when the order matters and an attribute set as a convenience for users whose setup does not require a certain order.

I believe I originally had it as a `listOf` but got complaints that it was cumbersome for simple setups.

I don't particularly mind switching back to `listOf` since I, in principle, agree about the rationale for deprecating `loaOf` but thought I'd make a note of this use-case.

File: lib/types.nix
Comment (rnhmjoj) [reply lib/types.nix]: I took me a while but I think this should do it:
```
trace: warning: In file <nixpkgs/nixos/modules/services/misc/nix-daemon.nix>
a list is being assigned to the option config.users.users.
This will soon be an error as type loaOf is deprecated.
See https://git.io/fj2zm for more information.
Do
  users.users =
    { nixbld1 = {...}; nixbld2 = {...}; nixbld3 = {...}; ...}
instead of
  users.users =
    [ { name = "nixbld1"; ...} { name = "nixbld2"; ...} { name = "nixbld3"; ...} ...]
```

**Comment by rnhmjoj** [comment]
@rycee I don't know how home-manager works: are you somehow bound to nixos option types? Couldn't you simply define an equivalent loaOf for home-manager?

File: lib/types.nix
Comment (rnhmjoj) [reply lib/types.nix]: Uhm, no it doesn't work when the set is missing the `name` attribute.

**Comment by alexarice** [comment]
Perhaps a solution is to separate off some "easy" examples, and get these merged progressively so that you don't need to do as much work fixing merge errors

**Comment by alexarice** [comment]
Had a little look at the changes, I wonder if while doing this it is worth changing things from `x.source = pkgs.writetext y z` to `x.text = z`

**Comment by rnhmjoj** [comment]
@alexarice: good idea

**Comment by rnhmjoj** [comment]
I have been running all the NixOS tests, it took a while because I couldn't figure out hydra and I had to use nix-build. Of all the tests in nixos/tests/all-tests.nix the only failures were:

|  package            |  cause
|:-------------------:|:--------------------
| beegfs              | unfree license
| buildbot            | **unknown**
| ceph                | broken package
| graphite            | insecure package
| hocker-fetchdocker  | broken dependency
| logstash            | **unknown**
| mathics             | broken package
| osquery             | broken package
| paperless           | timed out, **unknown**
| powerdns            | package tests failure
| taskserver          | **unknown**
| tinydns             | **unknown**
| xmonad              | timed out, **unknown**

The **unknown**s should be investigated.

File: lib/types.nix
Comment (rnhmjoj) [reply lib/types.nix]: I mean, it does but it's not pretty.

**Comment by rnhmjoj** [comment]
@alexarice: I have done the source -> text replacement on the files I have had already touched.

**Comment by alexarice** [comment]
I still think it could be easier to split this up slightly and merge in stages, it seems like an awful lot to test and commit in one go

**Comment by rnhmjoj** [comment]
It sure is but I'm not sure how to split it.
The deprecation warning could be merged by itself but it will cause a thousand of warnings. It will probably force people to rewrite their services creating even more conflicts to fix.
Maybe I could split d386bc3 into environment and users.

**Comment by alexarice** [comment]
You could make changes to modules before actually deprecating the type

File: lib/types.nix
Comment (Infinisil) [inline lib/types.nix]: Is this url persistent? Its shortness worries me

Comment (rnhmjoj) [reply lib/types.nix]: I have no idea what's the policy for the github url shortner but I yet to see one expire.

**Comment by rnhmjoj** [comment]
I've sorted out all the issue introduced with these changes, at least no test is failing anymore and nixpkgs evaluates. So... it should be ready. Someone else should test it, though.

**Comment by rnhmjoj** [comment]
I have found and fixed some small mistakes: It definitely needs more eyes.

**Comment by aanderse** [comment]
@rnhmjoj at this point you're just looking for people to review and test this to find any bugs that creep up?

**Comment by rnhmjoj** [comment]
@aanderse Yes, basically that.

**Comment by rnhmjoj** [comment]
All right, I should have fixed the eval now.

**Comment by aanderse** [comment]
@disassembler @worldofpeace as the `20.03` release managers can we possibly get some direction, or at least blessing on merging this PR?

If @rnhmjoj is able to rebase this and resolve all merge conflicts should we merge this and deal with issues as they arise? After 6 months of @rnhmjoj seeking feedback/testing/review on this I'm not sure how much more feedback we're going to receive.

Any input from release managers, especially on a PR of this scope, is always greatly valued.

**Comment by worldofpeace** [comment]
> @disassembler @worldofpeace as the `20.03` release managers can we possibly get some direction, or at least blessing on merging this PR?
> 
> If @rnhmjoj is able to rebase this and resolve all merge conflicts should we merge this and deal with issues as they arise? After 6 months of @rnhmjoj seeking feedback/testing/review on this I'm not sure how much more feedback we're going to receive.
> 
> Any input from release managers, especially on a PR of this scope, is always greatly valued.

I have no reason however to believe this PR could cause such a serious issue because I'm not aware of its exact changes.

I believe this change would have went in better if it wasn't in one go catching as many uses at once.
But that's still **greatly** appreciated, that takes a lot of effort. I do understand certain changes go under the radar without the right people for review or feedback. We only just merged https://github.com/NixOS/nixpkgs/pull/53843, and that was purposed at the beginning of this year. So similarly, something like this shouldn't go in without a review from at least a few people. I can find some time look at the changes, I do believe this belongs in 20.03.

**Comment by rnhmjoj** [comment]
@worldofpeace: Do you think I should write something about this in the release notes?

I would try running a final eval/all nixos tests before removing the WIP status but I can't get `hydra-eval-job` working in nixpkgs...

**Comment by aanderse** [comment]
@rnhmjoj I would suggest this deserves a mention in the release notes as it solves a problem more than one person has had and advertises the great work you've done! :tada:

File: lib/types.nix
Comment (worldofpeace) [reply lib/types.nix]: I don't believe github short urls expire, the only thing that would null it would be its nonexistence in the future.

**Review by infinisil — APPROVED** [review]
@rnhmjoj I don't think running all the tests is necessary, we'll see the failures in hydra over time (this is what it's there for after all :))

Other than having a release note entry this looks good to me

**Review by worldofpeace — APPROVED** [review]
Force push was just conflict resolution.

@rnhmjoj I believe this change is good, as mentioned we need a release note.
If docbook inconveniences you, I can translate your markups if that would be convenient for you.

If you could make the note similar to how our warning works in `types.nix`, explains how and why `loaOf` is deprecated and how code should be changed to not use it. And what phase of deprecation we're in, I believe it's not removed and code that uses it should work as it has.

**Comment by worldofpeace** [comment]
I actually believe it's unreasonable of us to stall merge on a release note.
If one is written in any form I can be available to commit that @rnhmjoj

**Comment by aanderse** [comment]
Thank you everyone, especially @rnhmjoj for your hard work!! :tada:

**Comment by rnhmjoj** [comment]
> If docbook inconveniences you, I can translate your markups if that would be convenient for you.

No, it's not a problem: I have written documentation for NixOS before.

> If you could make the note similar to how our warning works in types.nix, explains how and why loaOf is deprecated and how code should be changed to not use it. And what phase of deprecation we're in, I believe it's not removed and code that uses it should work as it has.

Ok, I'll do something like it. Thank you.

**Comment by petabyteboy** [comment]
I think there should be a pinned issue for this, since it will throw warnings and errors for many people and the efforts to fix those should be coordinated somehow.

**Comment by worldofpeace** [comment]
> I think there should be a pinned issue for this, since it will throw warnings and errors for many people and the efforts to fix those should be coordinated somehow.

Sounds good to me, will open.

**Comment by worldofpeace** [comment]
Made a little pop up issue https://github.com/NixOS/nixpkgs/issues/77189. Should have relevant strings within the body to pop up in relevancy searches.

**Comment by jtojnar** [comment]
The deprecation should probably be documented in `nixos/doc/manual/development/option-types.xml`.

**Comment by grahamc** [comment]
I just want to say I got the most wonderful error message from a Nix evaluation today:

    trace: warning: In file [...]/packet-spot-buildkite-agent/network.nix
    a list is being assigned to the option config.programs.ssh.knownHosts.
    This will soon be an error as type loaOf is deprecated.
    See https://git.io/fj2zm for more information.
    Do
      programs.ssh.knownHosts =
        { flexo.gsc.io = {...}; }
    instead of
      programs.ssh.knownHosts =
        [ { hostNames = [ "flexo.gsc.io" "r13y.com" "147.75.105.137" ]; ...} ]

This error message:

✔️ is specific: "a list is being assigned to the option config.programs.ssh.knownHosts"

✔️ tells me where to go for more information: https://git.io/fj2zm

✔️ tells me where the violation is: `.../packet-spot-buildkite-agent/network.nix`

✔️ **SHOWS** me the code which is wrong:

          programs.ssh.knownHosts =
            [ { hostNames = [ "flexo.gsc.io" "r13y.com" "147.75.105.137" ]; ...} ]
            
✔️ ***AND** tells me the solution*:

          programs.ssh.knownHosts =
            { flexo.gsc.io = {...}; }


This is incredible, and ticks the main three boxes of good errors:

1. What is the problem?
2. Why did it happen?
3. How do I fix it?

Way to go everyone, this is a massive step forward for our community and I am completely floored and delighted. Thank you.

**Pull request merged**
